### PR TITLE
Implement self updater

### PR DESCRIPTION
**Pull request by shirt-dev**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/pukkandan/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fixes the built in updater.

### CODE REVIEW

**Comment by pukkandan** [comment]
I am merging this for the time-being. But there is a small issue with the windows updater. If the file handle is not closed for whatever reason, the update won't be successful, but `Updated youtube-dlc to version %s` message is still issued. You can do `youtube-dlc.exe -U URL` to reproduce the issue reliably

Also, it would be nice to verify the downloaded file's checksum before replacing the running executable to ensure that the file has not been corrupted during download

**Comment by shirt-dev** [comment]
For the file handle issue, I believe this happens in normal youtube-dl as well.

As for checksum verification, if you modify the Github Actions workflows to make the releases have checksums for all of the binaries, I could implement this.

File: devscripts/pyinst.py [hunk 1]
```diff
@@ -27,7 +27,7 @@
 exec(compile(open('youtube_dlc/version.py').read(), 'youtube_dlc/version.py', 'exec'))
 VERSION = locals()['__version__']
 
-VERSION_LIST = VERSION.replace('-', '.').split('.')
```
Comment (shirt-dev) [inline devscripts/pyinst.py:30]: Why was this changed? This is for if there are two releases in 1 day.

File: devscripts/update-version.py [hunk 2]
```diff
@@ -8,7 +8,7 @@
 exec(compile(open('youtube_dlc/version.py').read(), 'youtube_dlc/version.py', 'exec'))
 old_version = locals()['__version__']
 
-old_version_list = old_version.replace('-', '.').split(".", 4)
```
Comment (shirt-dev) [inline devscripts/update-version.py:11]: Same as earlier, why was this changed?

**Comment by pukkandan** [comment]
> For the file handle issue, I believe this happens in normal youtube-dl as well.

Yes, which is why I merged despite the issue

> As for checksum verification, if you modify the Github Actions workflows to make the releases have checksums for all of the binaries, I could implement this.

Not now. We'll do it later

File: devscripts/pyinst.py (hunk 1 above)
Comment (pukkandan) [reply devscripts/pyinst.py:30]: In youtube-dlc, the version number was of the form `2021.02.09-1`. But I changed it to `2021.02.09.1`.

`.replace('-', '.')` was needed during the transition so that the new code can parse the old version number. But it is no longer needed
